include/zephyr/drivers/crc.h

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CRC_H
+#define ZEPHYR_INCLUDE_DRIVERS_CRC_H
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief CRC driver APIs
+ * @defgroup crc_interface CRC driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#define CRC_FLAG_REVERSE_INPUT  BIT(0)
+#define CRC_FLAG_REVERSE_OUTPUT BIT(1)
+
+#define CRC4_INIT_VAL        0x0
+#define CRC4_TI_INIT_VAL     0x0
+#define CRC7_BE_INIT_VAL     0x0
+#define CRC8_INIT_VAL        0x0
+#define CRC8_CCITT_INIT_VAL  0x00
+#define CRC16_INIT_VAL       0x0
+#define CRC16_ANSI_INIT_VAL  0x0
+#define CRC16_CCITT_INIT_VAL 0x0000
+#define CRC16_ITU_T_INIT_VAL 0x0000
+#define CRC24_PGP_INIT_VAL   0x0
+#define CRC32_C_INIT_VAL     0x00000000
+#define CRC32_IEEE_INIT_VAL  0xFFFFFFFFU
+
+/**
+ * @brief CRC algorithm enumeration
+ */
+enum crc_type {
+	CRC4,
+	CRC4_TI,
+	CRC7_BE,
+	CRC8,
+	CRC8_CCITT,
+	CRC16,
+	CRC16_ANSI,
+	CRC16_CCITT,
+	CRC16_ITU_T,
+	CRC24_PGP,
+	CRC32_C,
+	CRC32_IEEE,
+};
+
+/**
+ * @brief CRC state enumeration
+ */
+enum crc_state {
+	CRC_STATE_IDLE,
+	CRC_STATE_IN_PROGRESS,
+	CRC_STATE_COMPLETE
+};
+
+/**
+ * @brief Runtime driver configuration
+ * @param mutex Zephyr mutex object to make driver thread-safe when driver
+ *		is used within multiple threads
+ */
+struct crc_data {
+	struct k_mutex mutex;
+	k_timeout_t timeout;
+};
+
+typedef uint64_t crc_polynomial_t;
+typedef uint64_t crc_result_t;
+
+/**
+ * @brief CRC context structure
+ */
+struct crc_ctx {
+	enum crc_type type;          /**< Type of CRC algorithm */
+	enum crc_state state;        /**< State of CRC calculation */
+	const void *initial_value;   /**< Initial value for CRC calculation */
+	uint32_t flags;              /**< Flags for CRC calculation */
+	crc_polynomial_t polynomial; /**< Polynomial used in CRC calculation */
+	crc_result_t result;         /**< Store the CRC result */
+};
+
+typedef int (*crc_api_lock)(const struct device *dev, k_timeout_t timeout);
+typedef int (*crc_api_unlock)(const struct device *dev);
```

Comment:
These are out-of-scope imho: If driver wants to lock device when in use (i.e., `begin` called but `finish`/`verify`/whatever not called), it should be allowed to do so, but not forced to.